Mathlib/CategoryTheory/Galois/Decomposition.lean

```diff
@@ -160,6 +167,144 @@ lemma connected_component_unique {X A B : C} [IsConnected A] [IsConnected B] (a
+@[simp]
+private noncomputable def selfProd : C := ∏ (fun _ : F.obj X ↦ X)
+
+/- For `g : F.obj X → F.obj X`, this is the element in the fiber of the self product,
+which has at index `x : F.obj X` the element `g x`. -/
+private noncomputable def mkSelfProdFib : F.obj (selfProd F X) :=
+  (PreservesProduct.iso F _).inv ((Concrete.productEquiv (fun _ : F.obj X ↦ F.obj X)).symm id)
+
+@[simp]
+lemma mkSelfProdFib_map_π (t : F.obj X) : F.map (Pi.π _ t) (mkSelfProdFib F X) = t := by
+  rw [← congrFun (piComparison_comp_π F _ t), FintypeCat.comp_apply,
+    ← PreservesProduct.iso_hom]
+  simp only [mkSelfProdFib, FintypeCat.inv_hom_id_apply]
+  exact Concrete.productEquiv_symm_apply_π.{w, w, w+1} (fun _ : F.obj X ↦ F.obj X) id t
+
+variable {X} {A : C} [IsConnected A] (u : A ⟶ selfProd F X) [Mono u]
+  (a : F.obj A) (h : F.map u a = mkSelfProdFib F X) {F}
+
+/- For each `x : F.obj X`, this is the composition of `u` with the projection at `x`. -/
+@[simp]
+private noncomputable def selfProdProj (x : F.obj X) : A ⟶ X := u ≫ Pi.π _ x
+
+variable {u a}
+
+private lemma selfProdProj_fiber (x : F.obj X) :
+    F.map (selfProdProj u x) a = x := by
+  simp only [selfProdProj, selfProd, F.map_comp, FintypeCat.comp_apply, h]
+  rw [mkSelfProdFib_map_π F X x]
+
+/- An element `b : F.obj A` defines a permutation of the fiber of `X` by projecting onto the
+`F.map u b` factor. -/
+private noncomputable def fiberPerm (b : F.obj A) : F.obj X ≃ F.obj X := by
+  let σ (t : F.obj X) : F.obj X := F.map (selfProdProj u t) b
+  apply Equiv.ofBijective σ
+  apply Finite.injective_iff_bijective.mp
+  intro t s (hs : F.map (selfProdProj u t) b = F.map (selfProdProj u s) b)
+  show id t = id s
+  have h' : selfProdProj u t = selfProdProj u s := evaluationInjective_of_isConnected F A X b hs
+  rw [← selfProdProj_fiber h s, ← selfProdProj_fiber h t, h']
+
+/- Twisting `u` by `fiberPerm h b` yields an inclusion of `A` into `selfProd F X`. -/
+private noncomputable def selfProdPermIncl (b : F.obj A) : A ⟶ selfProd F X :=
+  u ≫ (Pi.whiskerEquiv (fiberPerm h b) (fun _ => Iso.refl X)).inv
+
+private instance (b : F.obj A) : Mono (selfProdPermIncl h b) := mono_comp _ _
+
+/- Key technical lemma: the twisted inclusion `selfProdPermIncl h b` maps `a` to `F.map u b`. -/
+private lemma selfProdTermIncl_fib_eq (b : F.obj A) :
+    F.map u b = F.map (selfProdPermIncl h b) a := by
+  apply Concrete.Pi.map_ext _ F
+  intro (t : F.obj X)
+  convert_to F.map (selfProdProj u t) b = _
+  · simp only [selfProdProj, map_comp, FintypeCat.comp_apply]; rfl
+  · dsimp only [selfProdPermIncl, Pi.whiskerEquiv]
+    rw [map_comp, FintypeCat.comp_apply, h]
+    convert_to F.map (selfProdProj u t) b =
+      (F.map (Pi.map' (fiberPerm h b) fun _ ↦ 𝟙 X) ≫
+      F.map (Pi.π (fun _ ↦ X) t)) (mkSelfProdFib F X)
+    rw [← map_comp, Pi.map'_comp_π, Category.comp_id, mkSelfProdFib_map_π F X (fiberPerm h b t)]
+    rfl
+
+/- `fiso` is obtained by considering `u` and `selfProdPermIncl h b`. Both
```

Comment:
What does `fiso` stand for?

Reply:
Nothing :D I changed the docstring to make it clearer.